src/DataStructures/Tensor/Expressions/Contract.hpp

```diff
@@ -175,56 +175,50 @@ struct TensorContract
 /*!
  * \ingroup TensorExpressionsGroup
  */
-template <typename ReplacedArg1, typename ReplacedArg2, typename T, typename X,
-          typename Symm, typename IndexList, typename Args>
-SPECTRE_ALWAYS_INLINE auto contract(
-    const TensorExpression<T, X, Symm, IndexList, Args>& t) {
-  return TensorContract<ReplacedArg1, ReplacedArg2, T, X, Symm, IndexList,
-                        Args>(~t);
-}
-
-namespace detail {
-// Helper struct to allow contractions by using repeated indices in operator()
-// calls to tensor.
-template <template <typename> class TE, typename ReplacedArgList, size_t I,
-          size_t TotalContracted, typename T>
-SPECTRE_ALWAYS_INLINE static constexpr auto fully_contract(const T& t) {
-  // TensorIndex types of pair of indices to contract
-  using lower_tensorindex = std::conditional_t<
-      std::is_same_v<
-          tmpl::index_of<ReplacedArgList, ti_contracted_t<I, UpLo::Lo, true>>,
-          tmpl::no_such_type_>,
-      ti_contracted_t<I, UpLo::Lo, false>, ti_contracted_t<I, UpLo::Lo, true>>;
-  using upper_tensorindex = std::conditional_t<
-      std::is_same_v<
-          tmpl::index_of<ReplacedArgList, ti_contracted_t<I, UpLo::Up, true>>,
-          tmpl::no_such_type_>,
-      ti_contracted_t<I, UpLo::Up, false>, ti_contracted_t<I, UpLo::Up, true>>;
-
-  // "first" and "second" refer to the position of the indices to contract
-  // in the list of generic indices, with "first" denoting leftmost
-  //
-  // e.g. `R(ti_A, ti_b, ti_a)` :
-  // - `first_replaced_tensorindex` refers to the contracted TensorIndex type
-  // for `ti_A`
-  // - `second_replaced_tensorindex` refers to the contracted TensorIndex type
-  // for `ti_a`
-  using first_replaced_tensorindex = tmpl::conditional_t<
-      (tmpl::index_of<ReplacedArgList, lower_tensorindex>::value <
-       tmpl::index_of<ReplacedArgList, upper_tensorindex>::value),
-      lower_tensorindex, upper_tensorindex>;
-  using second_replaced_tensorindex = tmpl::conditional_t<
-      (tmpl::index_of<ReplacedArgList, lower_tensorindex>::value <
-       tmpl::index_of<ReplacedArgList, upper_tensorindex>::value),
-      upper_tensorindex, lower_tensorindex>;
-
-  if constexpr (I == TotalContracted) {
-    return contract<first_replaced_tensorindex, second_replaced_tensorindex>(
-        TE<ReplacedArgList>(t));
+template <typename RepeatedTensorIndexValueList, typename T, typename X,
+          typename Symm, typename IndexList, typename ReplacedTensorIndexList>
+SPECTRE_ALWAYS_INLINE static constexpr auto contract(
+    const TensorExpression<T, X, Symm, IndexList, ReplacedTensorIndexList>&
+        t) noexcept {
+  if constexpr (tmpl::size<RepeatedTensorIndexValueList>::value == 0) {
+    // There aren't any repeated indices, so we just return the input
+    return ~t;
```

Comment:
why ~t instead of `t`? I'm not saying this is incorrect, just trying to understand :)

Reply:
This was discussed outside of GitHub, but summarizing here for documenting the reason:

Let's say we have a rank 2 `Tensor` named `R` and the expression, `R(ti_I, ti_i)`. When `contract` is called for the first time, the `else` case will be entered, which calls `contract` again and passes to it a `TensorContract` object. Then, the `if` case is entered in this second call. `t` is a `TensorContract` here, but without the `~`, if `t` is a `TensorContract`, then the return type will be its parent class, `TensorExpression<TensorContract<.., Tensor, ...>...>`. This is a problem because when we recurse down and call `TensorExpression::get` (multi-index version) to grab values from the original `Tensor` (`R`), we run into an issue where that expression's `t_` refers to itself instead of the original `Tensor`. This is because a `TensorContract` _is_ a `TensorExpression` of itself. The parent class should never be instantiated or used for this reason.

Instead, we need to downcast with `~t` here to ensure that a returned `TensorContract` is interpreted as a `TensorContract` instead of a `TensorExpression` of itself.